In [30]:
import pandas as pd
import numpy as np

results = pd.read_csv('../data/results.csv')

results.head()

FileNotFoundError: [Errno 2] No such file or directory: '../data/results.csv'

In [2]:
unique_stadiums = results['stadium'].value_counts()

unique_stadiums.head()

stadium
Twickenham             234
Lansdowne Road         229
Murrayfield Stadium    176
Millennium Stadium     134
National Stadium       116
Name: count, dtype: int64


Found 237 unique stadiums.

Twickenham, Lansdowne Road, and Murrayfield are among the most frequently used.

Not part of the overall project goals - I just wanted to know.

In [3]:
results['score_difference'] = results['home_score'] - results['away_score']

results.sort_values(
    by = 'score_difference',
    ascending = False
).head()

date     home_team  away_team  home_score  away_score  \
1566  1999-06-19  South Africa      Italy         101           0   
1585  1999-10-14   New Zealand      Italy         101           3   
1480  1997-06-21   New Zealand  Argentina          93           8   
1529  1998-06-27  South Africa      Wales          96          13   
2745  2023-09-29   New Zealand      Italy          96          17   

                                       competition             stadium  \
1566   1999 Italy rugby union tour of South Africa  Kings Park Stadium   
1585                          1999 Rugby World Cup    McAlpine Stadium   
1480            1997 Argentina tour of New Zealand       Athletic Park   
1529  1998 Wales tour of Zimbabwe and South Africa     Loftus Versfeld   
2745                          2023 Rugby World Cup      Parc Olympique   

              city       country  neutral  world_cup  score_difference  
1566        Durban  South Africa    False      False               101  
1585  Huddersfield       England     True       True                98  
1480    Wellington   New Zealand    False      False                85  
1529      Pretoria  South Africa    False      False                83  
2745          Lyon        France     True       True                79

Created a new column: score_difference = home_score - away_score

Sorting the results DataFrame by the score_difference column with the highest **home win** score difference (i.e, a positive score differential) by using ascending = False.

Two teams (South Africa and New Zealand) take up the top 5 biggest homes. 

Within the top 5 of this DataFrame sort, all of the home teams are Southern Hemimsphere nations and all the away teams are Northern Hemisphere nataions (Italy being on the wrong end of the two largest ever defeats).

I wonder if this reflects the dominance of the SOuthern Hemisphere over the Northern Hemisphere teams, or are just randonm chance within a very large data set.

Check dtypes attribute of the results DataFrame to see if date column is text/object...

It is.

In [4]:
results.dtypes

date                object
home_team           object
away_team           object
home_score           int64
away_score           int64
competition         object
stadium             object
city                object
country             object
neutral               bool
world_cup             bool
score_difference     int64
dtype: object

In [5]:
%cd ..
from scripts.data_load import date_to_year

results = date_to_year(results)

results.head()

/Users/douglasfairfield/Coding Projects/Data Projects/rugby_union_internationals


date home_team away_team  home_score  away_score  \
0  1871-03-27  Scotland   England           1           0   
1  1872-02-05   England  Scotland           2           1   
2  1873-03-03  Scotland   England           0           0   
3  1874-02-23   England  Scotland           1           0   
4  1875-02-15   England   Ireland           2           0   

                                competition                stadium       city  \
0     1871 Scotland v England International          Raeburn Place  Edinburgh   
1        1871–72 Home Nations International               The Oval     London   
2        1872–73 Home Nations International  West of Scotland F.C.    Glasgow   
3        1873–74 Home Nations International               The Oval     London   
4  1874–75 Home Nations rugby union matches               The Oval     London   

    country  neutral  world_cup  score_difference  year  
0  Scotland    False      False                 1  1871  
1   England    False      False                 1  1872  
2  Scotland    False      False                 0  1873  
3   England    False      False                 1  1874  
4   England    False      False                 2  1875

import the custom date_to_year function from the scripts module.

This fucntion creates a new 'year' colum that is the first four characters of the date column (i.e. the year formated as YYYY).

date_to_year function is applies the results DataFrame and first 5 rows are previewed to check it has worked (it has).

The import uses the Jupyter "magic" command so that Python can find the scripts folder at the same directory level

In [6]:
results['total_points'] = results['home_score'] + results['away_score']

results.head()

date home_team away_team  home_score  away_score  \
0  1871-03-27  Scotland   England           1           0   
1  1872-02-05   England  Scotland           2           1   
2  1873-03-03  Scotland   England           0           0   
3  1874-02-23   England  Scotland           1           0   
4  1875-02-15   England   Ireland           2           0   

                                competition                stadium       city  \
0     1871 Scotland v England International          Raeburn Place  Edinburgh   
1        1871–72 Home Nations International               The Oval     London   
2        1872–73 Home Nations International  West of Scotland F.C.    Glasgow   
3        1873–74 Home Nations International               The Oval     London   
4  1874–75 Home Nations rugby union matches               The Oval     London   

    country  neutral  world_cup  score_difference  year  total_points  
0  Scotland    False      False                 1  1871             1  
1   England    False      False                 1  1872             3  
2  Scotland    False      False                 0  1873             0  
3   England    False      False                 1  1874             1  
4   England    False      False                 2  1875             2

Created a new column: total_points = results['home_score'] + results['away_score']

The first 5 rows are previewed to check it has worked (it has).


In [7]:
results['year'] = results['year'].astype(int)

results.dtypes


date                object
home_team           object
away_team           object
home_score           int64
away_score           int64
competition         object
stadium             object
city                object
country             object
neutral               bool
world_cup             bool
score_difference     int64
year                 int64
total_points         int64
dtype: object

The 'year' column datatype is changed to an integer to allow for filtering by year later on.

I could probably (definitely) have incorporated this into the date_to_year custome function but I didn't think to at the time. 

I'll remember for next time.

In [8]:
results = results[['year', 'home_team', 'away_team', 'home_score', 'away_score', 'score_difference', 'total_points']]
results.head()

year home_team away_team  home_score  away_score  score_difference  \
0  1871  Scotland   England           1           0                 1   
1  1872   England  Scotland           2           1                 1   
2  1873  Scotland   England           0           0                 0   
3  1874   England  Scotland           1           0                 1   
4  1875   England   Ireland           2           0                 2   

   total_points  
0             1  
1             3  
2             0  
3             1  
4             2

I created a new DataFrame that only contained the columns I wanted.

I should probably have called it something other than "results" as this overrides the original results DataFrame.

**IMPORTANT NOTE:** I know I could easily change this but I want a record of my mistakes and things to improve on for future reference.

In [9]:
# No points for a try
# Team could only win a match if they successfully kicked a goal, either from a try or a drop goal

era_one = results[results['year'] < 1890]
era_one.head()

year home_team away_team  home_score  away_score  score_difference  \
0  1871  Scotland   England           1           0                 1   
1  1872   England  Scotland           2           1                 1   
2  1873  Scotland   England           0           0                 0   
3  1874   England  Scotland           1           0                 1   
4  1875   England   Ireland           2           0                 2   

   total_points  
0             1  
1             3  
2             0  
3             1  
4             2

Created a DataFrame for the pre-1890s era and called it era_one.
(This is why I needed to change the year datatype from object to int.)

The boolean mask and the filtering are done on a single line of code as it is a very simple filter and so still quite readable.


In [29]:
# Introduction of points for tries
# A try was worth 1 point, a conversion 2 points, a penalty goal 3 points, and a drop goal 4 points

results_after_1890_mask = results['year'] > 1890
results_before_1905_mask = results['year'] < 1905

era_two = results_after_1890_mask & results_before_1905_mask

era_two = results[era_two]

era_two.head()

year home_team away_team  home_score  away_score  score_difference  \
65  1891     Wales   England           3           7                -4   
66  1891  Scotland     Wales          15           0                15   
67  1891   Ireland   England           0           9                -9   
68  1891   Ireland  Scotland           0          14               -14   
69  1891     Wales   Ireland           6           4                 2   

    total_points  
65            10  
66            15  
67             9  
68            14  
69            10

Created a DataFrame for the 1890-1905 era and called it era_two.

I create two boolean masks and user the pandas & operator to combine the masks before applying the combined mask to the results DataFrame.

In [11]:
# 1905: A try increased to 3 points.

results_from_1905_mask = results['year'] >= 1905
results_before_1948_mask = results['year'] < 1948

era_three = results_from_1905_mask & results_before_1948_mask

era_three = results[era_three]

era_three.head()

year  home_team    away_team  home_score  away_score  score_difference  \
147  1905      Wales      England          25           0                25   
148  1905   Scotland        Wales           3           6                -3   
149  1905    Ireland      England          17           3                14   
150  1905   Scotland      Ireland           5          11                -6   
151  1905      Wales      Ireland          10           3                 7   
..    ...        ...          ...         ...         ...               ...   
491  1947  Australia  New Zealand           5          13                -8   
492  1947  Australia  New Zealand          14          27               -13   
493  1947   Scotland    Australia           7          16                -9   
494  1947    Ireland    Australia           3          16               -13   
495  1947      Wales    Australia           6           0                 6   

     total_points  
147            25  
148             9  
149            20  
150            16  
151            13  
..            ...  
491            18  
492            41  
493            23  
494            19  
495             6  

[349 rows x 7 columns]

In [12]:
# 1948: Drop goals were reduced from 4 points to 3 points, to ensure they didn’t outweigh tries

results_from_1948_mask = results['year'] >= 1948
results_before_1971_mask = results['year'] < 1971

era_four = results_from_1948_mask & results_before_1971_mask

era_four = results[era_four]

era_four.head()

year     home_team    away_team  home_score  away_score  \
496  1948        France      Ireland           6          13   
497  1948       England    Australia           0          11   
498  1948        France    Australia          13           6   
499  1948       England        Wales           3           3   
500  1948         Wales     Scotland          14           0   
..    ...           ...          ...         ...         ...   
880  1970     Australia     Scotland          23           3   
881  1970  South Africa  New Zealand          17           6   
882  1970  South Africa  New Zealand           8           9   
883  1970  South Africa  New Zealand          14           3   
884  1970  South Africa  New Zealand          20          17   

     score_difference  total_points  
496                -7            19  
497               -11            11  
498                 7            19  
499                 0             6  
500                14            14  
..                ...           ...  
880                20            26  
881                11            23  
882                -1            17  
883                11            17  
884                 3            37  

[389 rows x 7 columns]

In [31]:
# 1971: A try increased to 4 points to reward attacking rugby.

results_from_1971_mask = results['year'] >= 1971
results_before_1992_mask = results['year'] < 1992

era_five = results_from_1971_mask & results_before_1992_mask

era_five = results[era_five]

era_five.head()

year home_team away_team  home_score  away_score  score_difference  \
885  1971    France  Scotland          13           8                 5   
886  1971     Wales   England          22           6                16   
887  1971   Ireland    France           9           9                 0   
888  1971  Scotland     Wales          18          19                -1   
889  1971   Ireland   England           6           9                -3   

     total_points  
885            21  
886            28  
887            18  
888            37  
889            15

In [ ]:
# modern era: 1992: A try increased to 5 points to reward attacking rugby.
modern_era_mask = results['year'] >= 1992

modern_era = results[modern_era_mask]   

modern_era.head()

year home_team away_team  home_score  away_score  score_difference  \
1309  1992   Ireland     Wales          15          16                -1   
1310  1992  Scotland   England           7          25               -18   
1311  1992     Wales    France           9          12                -3   
1312  1992   England   Ireland          38           9                29   
1313  1992    France   England          13          31               -18   

      total_points  
1309            31  
1310            32  
1311            21  
1312            47  
1313            44

OK, so...given that I just did the same thing 6 times **AND** each time took several lines of code, I should 100% written a custom function in the scripts folder and used that. It would have save a lot of time and lines of code.

In [15]:
era_one_means = era_one[['score_difference', 'total_points']].mean().round(decimals=2)
era_two_means = era_two[['score_difference', 'total_points']].mean().round(decimals=2)
era_three_means = era_three[['score_difference', 'total_points']].mean().round(decimals=2)
era_four_means = era_four[['score_difference', 'total_points']].mean().round(decimals=2)
era_five_means = era_five[['score_difference', 'total_points']].mean().round(decimals=2) 
modren_era_means = modern_era[['score_difference', 'total_points']].mean().round(decimals=2)

era_means_tp_sd = pd.DataFrame(
    [era_one_means, era_two_means, era_three_means, era_four_means, era_five_means, modren_era_means],
    index = ['Pre-1890s', '1891-1905', '1905-1948', '1948-1971', '1971-1992', '1992-2024']
)

column_mapper = ['Mean Score Difference', 'Mean Total Points']

era_means_tp_sd.columns = column_mapper

era_means_tp_sd

Mean Score Difference  Mean Total Points
Pre-1890s                   0.41               1.69
1891-1905                   2.23              12.52
1905-1948                   1.00              19.12
1948-1971                   1.86              17.90
1971-1992                   3.94              31.95
1992-2024                   3.98              45.80

I created the mean for total_points and score_difference for every era DataFrame and then rounded the means to 2 dp.

I used these means create a new DataFrame using the pd.DataFrame method and set the index to be the eras as dates.

This gives me some intersting data. 

Let's defne an _“exciting”_ game of rugby as one that is **high-scoring** and reasonably close in score (**low score differential**)

Average total points per game have clearly risen
- Pre-1890s:  ~2 points         -> 0 points per try
- 1891–1905: ~12.5 points       -> try 1 point, con = 2, penalty = 3 points, drop goal = 4
- 1905–1948: ~19 points         -> try **3 points**, con = 2, penalty = 3 points, drop goal = 4
- 1948–1971: ~18 points         -> try 3 points, con = 2, penalty = 3 points, **drop goal = 3**
- 1991-1992: ~32 points         -> try **4 points**, con = 2, penalty = 3 points, drop goal = 3
- 1992–2024: ~46 points         -> try **5 point**, con = 2, penalty = 3 points, drop goal = 3

The average score difference has generally increased from pre-1890s to the modern era. 

1905-1948 saw a decrease in mean score difference from the previous era, while still seeing an increase in average total points per game. This makes it a potential candidate for the most exciting era. However, becuase the average score difference is less than the value of any score this may not have promoted attacking, try-focused, rugby since any score near the end of the game could flip the result.

Maybe I should look chamnge my definition of exciting where the score difference is less than a converted try but more than a penalty or drop goal? (although I do love a dramatic last gasp drop goal, but I'm a sample size of one so...others may disagree).

1948–1971 may have agreed with me since the points for a try remained the same but drop goal points were reduced from 4 to 3.

Both 1971-1992 and 1992-2024 have approximately 4-point score differences. This is a **one-score game** where a single unconverted try for the trailing team will flip the score. However, it is more than a penalty or a drop goal.

Maybe I should look at the percentage of games that are within unconverted try-flips vs drop-goal/penalty flips.

I could also look at the "most exciting" teams for each era and of all time?

In [32]:
# modern era games where the score difference was 4 and the total points were greater than the average total points scored in a game

score_difference_of_plus_four = modern_era['score_difference'] == 4 
score_difference_of_minus_four = modern_era['score_difference'] == -4
greater_than_average_points_mask = modern_era['total_points'] > modern_era['total_points'].mean()

combined_mask = (score_difference_of_plus_four | score_difference_of_minus_four) & greater_than_average_points_mask 

modern_era[combined_mask]

year     home_team     away_team  home_score  away_score  \
1498  1997        France  South Africa          32          36   
1509  1998         Italy      Scotland          25          21   
1535  1998   New Zealand     Australia          23          27   
1586  1999     Argentina       Ireland          28          24   
1618  2000     Australia   New Zealand          35          39   
1627  2000     Argentina  South Africa          33          37   
1676  2002         Wales        France          33          37   
1732  2003  South Africa      Scotland          29          25   
1742  2003  South Africa     Australia          26          22   
1800  2004     Australia  South Africa          30          26   
1845  2005   New Zealand  South Africa          31          27   
1872  2006       England       Ireland          24          28   
1999  2008     Australia   New Zealand          24          28   
2104  2010         Wales  South Africa          25          29   
2138  2011       Ireland        France          22          26   
2265  2013         Wales     Australia          26          30   
2372  2016         Wales      Scotland          27          23   
2379  2016       England         Wales          25          21   
2396  2016     Australia       England          40          44   
2440  2017     Argentina       England          34          38   
2566  2019     Australia         Wales          25          29   
2582  2020         Wales        France          23          27   
2623  2021        France      Scotland          23          27   
2699  2022        France  South Africa          30          26   
2731  2023      Scotland        France          25          21   
2749  2023       Ireland   New Zealand          24          28   

      score_difference  total_points  
1498                -4            68  
1509                 4            46  
1535                -4            50  
1586                 4            52  
1618                -4            74  
1627                -4            70  
1676                -4            70  
1732                 4            54  
1742                 4            48  
1800                 4            56  
1845                 4            58  
1872                -4            52  
1999                -4            52  
2104                -4            54  
2138                -4            48  
2265                -4            56  
2372                 4            50  
2379                 4            46  
2396                -4            84  
2440                -4            72  
2566                -4            54  
2582                -4            50  
2623                -4            50  
2699                 4            56  
2731                 4            46  
2749                -4            52

I used a boolean mask / filter to find all the games in the modern era that have a score difference of 4 or -4 (for away wins) and a total points value greater than the average of the modern era.

This  as this gave 16 results. 

I nearly forgot to include the -4 results...but I didn't.

I should work out which nation is the most exciting of the modern era (according to my definition) by counting which team appears the most in this data set (maybe when I result to this project in the future).

In [17]:
modern_era[combined_mask].score_difference.value_counts()

score_difference
-4    16
 4    10
Name: count, dtype: int64

In [18]:
modern_era_excietment_rate = (26 / 1474 * 100)
modern_era_excietment_rate

1.7639077340569878

In [19]:
# 1971-1992 era games where the score difference was 4 and the total points were greater than the average total points scored in a game

score_difference_less_than_three = era_five['score_difference'] < 3
score_difference_more_than_minus_three  = era_five['score_difference'] > -3
greater_than_average_points_mask = era_five['total_points'] > era_five['total_points'].mean()

combined_mask = score_difference_less_than_three & score_difference_more_than_minus_three & greater_than_average_points_mask 

era_five[combined_mask].info()

<class 'pandas.core.frame.DataFrame'>
Index: 21 entries, 888 to 1304
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   year              21 non-null     int64 
 1   home_team         21 non-null     object
 2   away_team         21 non-null     object
 3   home_score        21 non-null     int64 
 4   away_score        21 non-null     int64 
 5   score_difference  21 non-null     int64 
 6   total_points      21 non-null     int64 
dtypes: int64(5), object(2)
memory usage: 1.3+ KB


In [20]:
era_five_excietment_rate = (21 / 424 * 100)
era_five_excietment_rate

4.952830188679245

In [21]:
modern_era

year    home_team     away_team  home_score  away_score  \
1309  1992      Ireland         Wales          15          16   
1310  1992     Scotland       England           7          25   
1311  1992        Wales        France           9          12   
1312  1992      England       Ireland          38           9   
1313  1992       France       England          13          31   
...    ...          ...           ...         ...         ...   
2778  2024    Argentina        France          33          25   
2779  2024    Australia  South Africa           7          33   
2780  2024  New Zealand     Argentina          30          38   
2781  2024  New Zealand     Argentina          42          10   
2782  2024    Australia  South Africa          12          30   

      score_difference  total_points  
1309                -1            31  
1310               -18            32  
1311                -3            21  
1312                29            47  
1313               -18            44  
...                ...           ...  
2778                 8            58  
2779               -26            40  
2780                -8            68  
2781                32            52  
2782               -18            42  

[1474 rows x 7 columns]

In [22]:
modern_era['score_difference'].max()

np.int64(101)

In [23]:
modern_era['score_difference'].min()

np.int64(-64)

In [24]:
biggest_home_win = modern_era[modern_era['score_difference'] == 101]
biggest_home_win

year     home_team away_team  home_score  away_score  score_difference  \
1566  1999  South Africa     Italy         101           0               101   

      total_points  
1566           101

In [25]:
biggest_away_win = modern_era[modern_era['score_difference'] == -64]
biggest_away_win

year home_team    away_team  home_score  away_score  score_difference  \
1421  1995     Italy  New Zealand           6          70               -64   

      total_points  
1421            76

In [26]:
modern_era_nations = modern_era.groupby('home_team').agg(
    {
        'home_team': 'count',
        'score_difference': 'mean',
        'total_points': 'mean'
    }
)

modern_era_nations['score_difference'] = modern_era_nations['score_difference'].round(decimals=2)
modern_era_nations['total_points'] = modern_era_nations['total_points'].round(decimals=2)

modern_era_nations.sort_values(
    by = 'score_difference',
    ascending = False
)


home_team  score_difference  total_points
home_team                                              
New Zealand         164             19.29         50.25
South Africa        149             10.23         48.99
England             174              9.15         44.45
Australia           170              7.06         46.40
Ireland             145              6.92         41.89
France              159              5.32         44.52
Wales               164             -0.67         43.49
Scotland            134             -4.07         41.11
Argentina           109             -8.28         49.05
Italy               106            -17.92         49.18

In [27]:
# modern era average score_difference: 3.98 
# modern era average total points per game: 45.80

%cd ..
from scripts.data_load import group_by_year_mean_score_diff_mean_total_points, sd_tp_filter

modern_era_mean_sd_and_tp = group_by_year_mean_score_diff_mean_total_points(modern_era)

sd_tp_filter(modern_era_mean_sd_and_tp, 5, 3, 45.80)['year'].count()



/Users/douglasfairfield/Coding Projects/Data Projects


np.int64(12)

12 years in the modern era that are above average excitment

In [28]:
# era 5 average score_difference: 3.94 
# era 5 average total points per game: 31.95

era_five_mean_sd_and_tp = group_by_year_mean_score_diff_mean_total_points(era_five)

sd_tp_filter(era_five_mean_sd_and_tp, 4, 3, 31.95)['year'].count()

np.int64(8)

8 years in the modern era that are above average excitment